### ISSUE TITLE
time_synchronized() when using CPU for inference on a GPU enabled workstation?

### ISSUE THREAD

**Issue by tayyabmujahid** [ultralytics/yolov5#1816]
Im trying to measure time taken for inference using a CPU vs GPU.

I set --device as cpu when i run detect.py but the method time_synchronized() checks torch.cuda.is_available() which obviously is True as the gpu is available but not used.
Also i've noticed that when i comment `torch.cuda.synchronize() if torch.cuda.is_available() else None` (in time_synchronized() method)while using --device as cpu the inference speeds up.

Shouldn't the time_synchronize() be connected to the --device parameter?

**Comment by glenn-jocher** [comment]
@tayyabmujahid interesting observation. I suppose in theory yes, though this would also add complexity to the codebase. Have you quantified any effects from the synchronized times on --device CPU on a CUDA machine?

**Comment by tayyabmujahid** [comment]
@glenn-jocher following are the details

My workstation specs:
GPU: Quadro P600, 2GB
RAM: 32 GB
CPU:Intel Xeon

img_size = 640 x 480
using pretrained yolov5s on bus.jpg and zidane.jpg images respectively

1.  --device CPU  0.103s, 0.087s (total time after results are saved 2.212s)
2.  --device CPU (with commented [line 74](https://github.com/ultralytics/yolov5/blob/0b6266f5e0eab11218871d5560bf9b93f7547aac/utils/torch_utils.py#L74))- 0.109s, 0.091s (0.270s)
3. -- device GPU - 0.1615s per image - 0.072s,0.058s (0.305s)

though not the most comprehensive experiment, also i have to be more specific than my previous comment. 

a) The per image performance remain more or less the same on the contrary when [line 74 ](https://github.com/ultralytics/yolov5/blob/0b6266f5e0eab11218871d5560bf9b93f7547aac/utils/torch_utils.py#L74) is commented it performs little worse than than when it was not.
b) But the total time taken ( time taken for making the predictions and saving the images with bboxes) changes dramatically as noticed above
c) And i face gpu memory issues some times when using cpu for inference with [line74](https://github.com/ultralytics/yolov5/blob/0b6266f5e0eab11218871d5560bf9b93f7547aac/utils/torch_utils.py#L74) not commented.

**Comment by glenn-jocher** [comment]
Oh, so the main takeaway is that the per image speeds are unaffected but the overall execution time reported is much longer (2.2s vs 0.27s)?

I'm not sure what modifications to recommend. Have you tried setting CUDA_VISIBLE_DEVICES to '' in select_device() when a CPU is requested?

**Comment by glenn-jocher** [comment]
If I run on a CPU only machine (2020 iMac) I get the same result as your #1825 

```
Fusing layers... 
Model Summary: 232 layers, 7459581 parameters, 0 gradients, 17.5 GFLOPS
image 1/2 /Users/glennjocher/PycharmProjects/yolov5/data/images/bus.jpg: 640x480 4 persons, 1 buss, 1 skateboards, Done. (0.121s)
image 2/2 /Users/glennjocher/PycharmProjects/yolov5/data/images/zidane.jpg: 384x640 2 persons, 2 ties, Done. (0.106s)
Results saved to runs/detect/exp2
Done. (0.300s)
```

**Comment by glenn-jocher** [comment]
If I run in Colab with --device cpu I also observe the same large slowdown. Well that's very useful that you've isolated the slowdown to the torch.cuda.synchronize() call. Yes, I agree this is something we should fix. I'll look into it.

<img width="1211" alt="Screen Shot 2021-01-03 at 9 51 07 AM" src="https://user-images.githubusercontent.com/26833433/103485174-553ffa00-4da9-11eb-86ba-7a8cc5561f13.png">

**Comment by glenn-jocher** [comment]
@tayyabmujahid this issue should be resolved now in PR #1826. Please git pull to receive this update and let us know if you have any other issues!

**Comment by glenn-jocher** [comment]
<img width="1199" alt="Screen Shot 2021-01-03 at 11 16 58 AM" src="https://user-images.githubusercontent.com/26833433/103486815-7ad30080-4db5-11eb-917a-1c763da52103.png">
